File: shared/nodejs/pplibs/customErrors.js
var AWS = require('aws-sdk');
AWS.config.update({ region: 'us-west-1' });
let sns = new AWS.SNS();


function reportToRollbar(message, rollbarTopic, referer = "NO_REFERER") {
    return new Promise((resolve, reject) => {

        if (typeof message === "string") {
            message = new ProPairError(message)
        }
        if (typeof message.referer === 'undefined' || message.referer == null) {
            message.referer = referer
            message.name = referer
        }
        message.stack = message.stack.replace(/\"/g, "")
        message.message = message.message.replace(/\"/g, "")
        var errorParams = {
            Message: JSON.stringify(message, replaceErrors),
            Subject: "Rollbar Error",
            TopicArn: rollbarTopic
        };


        sns.publish(errorParams, function (err, data) {
            if (err) {
                console.log("ERROR: Failed to report exception to Rollbar");
                console.log(err);
                console.log(e);
                reject(err)
            } else {
                console.log("ERROR POSTED TO ROLLBAR");
                resolve()
            }
        });
    })
}


class GeneralProPairError extends Error {
    constructor(message) {
        super(message);
        this.name = this.constructor.name;
    }
}

function replaceErrors(key, value) {
    if (value instanceof GeneralProPairError) {
        var error = {};

        Object.getOwnPropertyNames(value).forEach(function (key) {
            error[key] = value[key];
        });

        return error;
    }

    return value;
}

class rollbarError extends GeneralProPairError {
    constructor(payload) {
        super(payload.message);
        this.extra = {}
        for (const key of Object.keys(payload)) {
            this.extra[key] = payload[key]
        }
        this.stack = payload.stack
    }
}

class MapError extends GeneralProPairError {

    static get map_error_types() {
        let types = [
            "Source_Missing",
            "Source_Detail_Missing",
            "All_Sources_Missing",
            "Source_Not_Found",
            "Campaign_Group_for_Source_Detail_and_Source_Not_Found",
            "Source_Detail_for_Source_Not_Found",
            "Mapping_Failure"
        ];
        return types
    }

    constructor(payload) {
        super(payload.message);
        if (typeof payload.lead !== 'undefined' && payload.lead != null) {
            this.lead = payload.lead
            this.customer = payload.customer
            this.mappingErrorType = MapError.map_error_types[payload.type]
        }
    }
}

class ProPairError extends GeneralProPairError {
    constructor(payload, type = null) {
        if (typeof payload === "string") {
            super(payload)
        }
        else {
            super(String(payload))
        }
        this.stack = payload.stack || this.stack

        if (typeof type !== "undefined" || type != null) {
            this.name = type
            this.referer = type
        }
    }
}

exports.reportToRollbar = reportToRollbar
exports.rollbarError = rollbarError
exports.MapError = MapError
exports.ProPairError = ProPairError
exports.replaceErrors = replaceErrors

File: shared/python3/python/pplibs/customErrors.py
import boto3
from json import dumps
import pickle
_SNS = boto3.client('sns')

class ProPairError(Exception):
    def __init__(self, err, error_type="NO_REFERER", exec_info="", stack=""):

        # Call the base class constructor with the parameters it needs
        super(ProPairError, self).__init__(err)

        if isinstance(err, str):
            self.message = err.replace("\"", "")
        else:
            self.message = str(err).replace("\"", "")

        if isinstance(exec_info, str):
            self.exec_info = exec_info.replace("\"", "'")
        else:
            self.exec_info = str(exec_info).replace("\"", "'")
        self.stack = stack.replace("\"", "")
        # Now for your custom code...
        self.referer = error_type
        self.name = error_type
        self.error_type = "PythonException"

def report_to_rollbar(topic, py_err):
    rollbar_error = {}
    if not isinstance(py_err, ProPairError):
        try:
            if isinstance(py_err, str):
                rollbar_error = json.dumps(py_err)
            else:
                if hasattr(py_err, "args"):
                    rollbar_error = py_err.__dict__
                    rollbar_error["message"] = py_err.args[0]
        except Exception as e:
            rollbar_error["message"] = "No error format"
    else:
        rollbar_error = py_err.__dict__
    rollbar_error["message"] = py_err.message if hasattr(py_err, "message") else ""
    rollbar_error = dumps(rollbar_error)
    response = _SNS.publish(
        TopicArn=topic, 
        Subject="Production Error in Recommend Lambda",   
        Message=rollbar_error
    )
    print("Response: {}".format(response))
